mountpoint-s3-fs/src/memory.rs

```diff
@@ -0,0 +1,457 @@
+use std::alloc::{self, Layout, handle_alloc_error};
+use std::io::Read;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::{Arc, Mutex, RwLock, Weak};
+use std::time::Duration;
+
+use bytes::Bytes;
+use mountpoint_s3_client::config::MemoryPool;
+
+/// A memory pool implementation that uses paged buffers.
+#[derive(Debug, Clone)]
+pub struct PagedPool {
+    inner: Arc<PagedPoolInner>,
+}
+
+impl PagedPool {
+    /// Creates a pool that reuses buffers of the given sizes.
+    pub fn new(buffer_sizes: impl Into<Vec<usize>>) -> Self {
+        let mut buffer_sizes: Vec<_> = buffer_sizes.into();
+        buffer_sizes.sort();
+        buffer_sizes.dedup();
+
+        let size_pools = buffer_sizes
+            .iter()
+            .map(|&buffer_size| SizePool {
+                pages: Default::default(),
+                empty_page_count: AtomicUsize::new(0),
+                buffer_size,
+            })
+            .collect();
+
+        let inner = PagedPoolInner { size_pools };
+        Self { inner: Arc::new(inner) }
+    }
+
+    /// Trim empty pages in the pool.
+    pub fn trim(&self) -> bool {
+        self.inner.trim()
+    }
+
+    /// Schedule recurring calls to [trim()](Self::trim()).
+    pub fn schedule_trim(&self, recurring_time: Duration) {
+        let weak = Arc::downgrade(&self.inner);
+        std::thread::spawn(move || {
+            loop {
+                std::thread::sleep(recurring_time);
+                let Some(pool) = weak.upgrade() else {
+                    return;
+                };
+                pool.trim();
+            }
+        });
+    }
+
+    /// Reads `size` bytes into a new buffer from the pool.
+    pub fn read_exact(&self, read: &mut impl Read, size: usize) -> std::io::Result<Bytes> {
+        let mut buffer = self.get_buffer(size);
+        read.read_exact(buffer.as_mut())?;
+        Ok(Bytes::from_owner(buffer))
+    }
+
+    fn get_buffer(&self, size: usize) -> PoolBuffer {
+        PoolBuffer(match self.reserve(size) {
+            Some(buffer_ptr) => {
+                metrics::histogram!("pool.reserve", "type" => "primary").record(size as f64);
```

Comment:
We should add a metric for 'wasted' bytes to help see about adding new buckets here as well